// @cmsgov/design-system-core/dist/components/MonthPicker/MonthPicker.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getMonthNames = getMonthNames;
exports.default = exports.MonthPicker = void 0;

var _Button = _interopRequireDefault(require("../Button/Button"));

var _Choice = _interopRequireDefault(require("../ChoiceList/Choice"));

var _FormLabel = _interopRequireDefault(require("../FormLabel/FormLabel"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireDefault(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

var NUM_MONTHS = 12;

var monthNumbers = function () {
  var months = [];

  for (var m = 1; m <= NUM_MONTHS; m++) {
    months.push(m);
  }

  return months;
}();

var MonthPicker = /*#__PURE__*/function (_React$PureComponent) {
  _inherits(MonthPicker, _React$PureComponent);

  function MonthPicker(props) {
    var _this;

    _classCallCheck(this, MonthPicker);

    _this = _possibleConstructorReturn(this, _getPrototypeOf(MonthPicker).call(this, props));
    _this.months = getMonthNames(props.locale);
    _this.monthsLong = getMonthNames(props.locale, false);

    if (typeof props.selectedMonths === 'undefined') {
      _this.isControlled = false; // Since this isn't a controlled component, we need a way
      // to track when the value has changed.

      _this.state = {
        selectedMonths: props.defaultSelectedMonths || []
      };
    } else {
      _this.isControlled = true;
    }

    return _this;
  }

  _createClass(MonthPicker, [{
    key: "selectedMonths",
    value: function selectedMonths() {
      if (this.isControlled) {
        return this.props.selectedMonths;
      }

      return this.state.selectedMonths;
    }
  }, {
    key: "disabledMonths",
    value: function disabledMonths() {
      return this.props.disabledMonths || [];
    }
  }, {
    key: "handleChange",
    value: function handleChange(event) {
      if (this.props.onChange) {
        this.props.onChange(event);
      }

      if (!this.isControlled) {
        var month = parseInt(event.target.value);
        var selectedMonths = this.state.selectedMonths.slice();

        if (selectedMonths.includes(month)) {
          selectedMonths.splice(selectedMonths.indexOf(month), 1);
        } else {
          selectedMonths.push(month);
        }

        this.setState({
          selectedMonths: selectedMonths
        });
      }
    }
  }, {
    key: "handleSelectAll",
    value: function handleSelectAll() {
      if (this.props.onSelectAll) {
        this.props.onSelectAll();
      }

      if (!this.isControlled) {
        var disabledMonths = this.disabledMonths();
        var selectedMonths = monthNumbers.filter(function (m) {
          return !disabledMonths.includes(m);
        });
        this.setState({
          selectedMonths: selectedMonths
        });
      }
    }
  }, {
    key: "handleClearAll",
    value: function handleClearAll() {
      if (this.props.onClearAll) {
        this.props.onClearAll();
      }

      if (!this.isControlled) {
        this.setState({
          selectedMonths: []
        });
      }
    }
  }, {
    key: "renderMonths",
    value: function renderMonths() {
      var _this2 = this;

      var selectedMonths = this.selectedMonths();
      var disabledMonths = this.disabledMonths();
      var _this$props = this.props,
          name = _this$props.name,
          inversed = _this$props.inversed;
      return _react.default.createElement("ol", {
        className: "ds-c-list--bare ds-u-display--flex ds-u-justify-content--between ds-u-flex-wrap--wrap"
      }, this.months.map(function (month, i) {
        return _react.default.createElement("li", {
          key: month
        }, _react.default.createElement(_Choice.default, {
          "aria-label": _this2.monthsLong[i],
          checked: selectedMonths.includes(i + 1),
          className: "ds-c-month-picker__month",
          disabled: disabledMonths.includes(i + 1),
          inversed: inversed,
          onChange: function onChange(e) {
            return _this2.handleChange(e);
          },
          name: name,
          value: i + 1
        }, month));
      }));
    }
  }, {
    key: "renderButton",
    value: function renderButton(text, pressed, onClick) {
      return _react.default.createElement(_Button.default, {
        "aria-pressed": pressed,
        size: "small",
        className: "ds-u-margin-right--1 ds-u-float--left",
        onClick: onClick,
        inversed: this.props.inversed,
        variation: this.props.buttonVariation
      }, text);
    }
  }, {
    key: "renderLabel",
    value: function renderLabel() {
      return _react.default.createElement(_FormLabel.default, {
        component: "legend",
        errorMessage: this.props.errorMessage,
        requirementLabel: this.props.requirementLabel,
        hint: this.props.hint,
        inversed: this.props.inversed
      }, this.props.label);
    }
  }, {
    key: "render",
    value: function render() {
      var _this3 = this;

      var _this$props2 = this.props,
          selectAllText = _this$props2.selectAllText,
          clearAllText = _this$props2.clearAllText;
      var selectedMonths = this.selectedMonths();
      var disabledMonths = this.disabledMonths();
      var selectAllPressed = selectedMonths.length === NUM_MONTHS - disabledMonths.length;
      var clearAllPressed = selectedMonths.length === 0;
      var classes = (0, _classnames.default)('ds-c-month-picker', 'ds-c-fieldset', 'ds-u-margin-y--3', this.props.className);
      return _react.default.createElement("div", {
        className: classes
      }, _react.default.createElement("fieldset", {
        className: "ds-c-fieldset"
      }, this.renderLabel(), _react.default.createElement("div", {
        className: "ds-c-month-picker__buttons ds-u-margin-top--2 ds-u-margin-bottom--1 ds-u-clearfix"
      }, this.renderButton(selectAllText, selectAllPressed, function () {
        return _this3.handleSelectAll();
      }), this.renderButton(clearAllText, clearAllPressed, function () {
        return _this3.handleClearAll();
      })), _react.default.createElement("div", {
        className: "ds-c-month-picker__months"
      }, this.renderMonths())));
    }
  }]);

  return MonthPicker;
}(_react.default.PureComponent);

exports.MonthPicker = MonthPicker;
MonthPicker.defaultProps = {
  selectAllText: 'Select all',
  clearAllText: 'Clear all'
};
MonthPicker.propTypes = {
  /**
   * The `input` field's `name` attribute
   */
  name: _propTypes.default.string.isRequired,

  /**
   * A [BCP 47 language tag](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl#Locale_identification_and_negotiation)
   * for month name localization. For example: Passing `es-US` as a value
   * will render month names in Spanish.
   */
  locale: _propTypes.default.string,

  /**
   * Additional classes to be added to the root element.
   */
  className: _propTypes.default.string,

  /**
   * Applies the "inverse" UI theme
   */
  inversed: _propTypes.default.bool,

  /**
   * Variation string to be applied to buttons. See [Button component]({{root}}/components/button/#components.button.react)
   */
  buttonVariation: _propTypes.default.string,

  /**
   * Label for the field
   */
  label: _propTypes.default.node.isRequired,
  errorMessage: _propTypes.default.node,

  /**
   * Additional hint text to display
   */
  hint: _propTypes.default.node,

  /**
   * Text showing the requirement ("Required", "Optional", etc.). See [Required and Optional Fields]({{root}}/guidelines/forms/#required-and-optional-fields).
   */
  requirementLabel: _propTypes.default.node,

  /**
   * Array of month numbers, where `1` is January, and any month included
   * is disabled for selection.
   */
  disabledMonths: _propTypes.default.arrayOf(_propTypes.default.number),

  /**
   * Array of month numbers, where `1` is January, and any month included
   * is selected. This will render a read-only field. If the field should
   * be mutable, use `defaultSelectedMonths`.
   */
  selectedMonths: _propTypes.default.arrayOf(_propTypes.default.number),

  /**
   * Array of month numbers, where `1` is January, and any month included
   * is selected by default. Sets the initial checked state for the 12 month
   * checkboxes. Use this for an uncontrolled component; otherwise, use the
   * `selectedMonths` property.
   */
  defaultSelectedMonths: _propTypes.default.arrayOf(_propTypes.default.number),

  /**
   * A callback function that's invoked when a month's checked state is changed.
   * Note: This callback is not called when a month is selected or deselected
   * via the "Select all" or "Clear all" buttons – use the `onSelectAll` and
   * `onClearAll` event handlers for those instances.
   */
  onChange: _propTypes.default.func,
  onSelectAll: _propTypes.default.func,
  onClearAll: _propTypes.default.func,

  /**
   * For internationalization purposes, the text for the "Select all"
   * button must be passed in as a prop.
   */
  selectAllText: _propTypes.default.string.isRequired,

  /**
   * For internationalization purposes, the text for the "Clear all"
   * button must be passed in as a prop.
   */
  clearAllText: _propTypes.default.string.isRequired
};
var _default = MonthPicker;
/**
 * Generates an array of month names according to the given or default locale
 *
 * @param  {string} [locale] locale for generating month names
 * @param  {boolean} [short] whether to return short month names
 * @return {string[]}        array of month names
 */

exports.default = _default;

function getMonthNames(locale) {
  var short = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : true;
  var options = {
    month: short ? 'short' : 'long'
  };
  var months = [];

  for (var i = 0; i < NUM_MONTHS; i++) {
    var date = new Date();
    date.setMonth(i, 1);
    months.push(date.toLocaleString(locale, options));
  }

  return months;
}
